File: src/apple/Packages/MapleUI/Sources/MapleUI/Molecules/MuiCommandMenu.swift
// MuiCommandMenu.swift — Maple UI Molecules-L1 (unified-component-
// catalog.md §2.4; Built from: Popover, Input, Icon, Text). A searchable
// command palette: an MuiInput filter on top of an anchored MuiPopover
// result list, filtering by substring match against each command's label.

import SwiftUI

public struct MuiCommandItem: Identifiable, Sendable {
    public let id: String
    public let label: String
    public let icon: String?
    public let shortcut: String?

    public init(id: String, label: String, icon: String? = nil, shortcut: String? = nil) {
        self.id = id
        self.label = label
        self.icon = icon
        self.shortcut = shortcut
    }
}

public struct MuiCommandMenu<Trigger: View>: View {
    @Binding public var open: Bool
    public let placement: MuiPopoverPlacement
    public let commands: [MuiCommandItem]
    public let placeholder: String
    public let select: (String) -> Void
    @ViewBuilder public let trigger: Trigger

    @State private var query = ""
    @State private var activeIndex = 0

    public init(
        open: Binding<Bool>,
        placement: MuiPopoverPlacement = .bottom,
        commands: [MuiCommandItem],
        placeholder: String = "Type a command",
        select: @escaping (String) -> Void,
        @ViewBuilder trigger: () -> Trigger
    ) {
        self._open = open
        self.placement = placement
        self.commands = commands
        self.placeholder = placeholder
        self.select = select
        self.trigger = trigger()
    }

    private var filtered: [MuiCommandItem] {
        let q = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !q.isEmpty else { return commands }
        return commands.filter { $0.label.lowercased().contains(q) }
    }

    private var clampedActiveIndex: Int {
        MuiMenuNavMath.clampedIndex(activeIndex, count: filtered.count)
    }

    public var body: some View {
        trigger
            .muiPopover(isPresented: open, placement: placement, closeRequested: { open = false }) {
                menu
            }
            // Every (re)open starts from a clean search, same as a real
            // command palette — the previous query never survives a close.
            .onChange(of: open) { _, isOpen in
                if isOpen {
                    query = ""
                    activeIndex = 0
                }
            }
    }

    private var menu: some View {
        VStack(alignment: .leading, spacing: MuiTokens.spacingXs) {
            MuiInput(value: $query, accessibilityLabel: placeholder, placeholder: placeholder, prefixIcon: "magnifyingglass")
                .onChange(of: query) { activeIndex = 0 }

            if filtered.isEmpty {
                MuiText("No commands found", variant: .body, color: .muted)
                    .padding(.horizontal, MuiTokens.spacingSm)
                    .padding(.vertical, MuiTokens.spacingXs)
            }

            ForEach(Array(filtered.enumerated()), id: \.element.id) { index, command in
                row(command, index: index)
            }
        }
        .frame(minWidth: 240)
        #if os(macOS)
        .focusable()
        .onKeyPress(.downArrow) { moveActive(1); return .handled }
        .onKeyPress(.upArrow) { moveActive(-1); return .handled }
        .onKeyPress(.return) { activateHighlighted(); return .handled }
        #endif
    }

    private func row(_ command: MuiCommandItem, index: Int) -> some View {
        Button {
            select(command.id)
            open = false
        } label: {
            HStack(spacing: MuiTokens.spacingSm) {
                if let icon = command.icon {
                    MuiIcon(name: icon, size: .sm, color: MuiTokens.textMuted)
                }
                MuiText(command.label, variant: .rowLabel)
                Spacer(minLength: MuiTokens.spacingSm)
                if let shortcut = command.shortcut {
                    MuiText(shortcut, variant: .toolLabel, color: .muted)
                }
            }
            .padding(.horizontal, MuiTokens.spacingSm)
            .padding(.vertical, MuiTokens.spacingXs)
            .frame(minHeight: 32)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                index == clampedActiveIndex ? MuiTokens.surfaceHover : .clear,
                in: RoundedRectangle(cornerRadius: MuiTokens.radiusSm, style: .continuous)
            )
        }
        .buttonStyle(.plain)
    }

    private func moveActive(_ direction: Int) {
        guard !filtered.isEmpty else { return }
        activeIndex = MuiMenuNavMath.wrappedIndex(current: clampedActiveIndex, direction: direction, count: filtered.count)
    }

    private func activateHighlighted() {
        let index = clampedActiveIndex
        guard filtered.indices.contains(index) else { return }
        select(filtered[index].id)
        open = false
    }
}

#Preview("MuiCommandMenu") {
    struct Demo: View {
        @State private var open = false

        var body: some View {
            MuiCommandMenu(
                open: $open,
                commands: [
                    MuiCommandItem(id: "export", label: "Export…", icon: "square.and.arrow.up", shortcut: "⌘E"),
                    MuiCommandItem(id: "rename", label: "Batch Rename…", icon: "pencil"),
                    MuiCommandItem(id: "crop", label: "Crop", icon: "crop", shortcut: "C"),
                ],
                select: { _ in }
            ) {
                MuiButton(label: "Command palette", variant: .secondary, leadingIcon: "command") { open.toggle() }
            }
        }
    }
    return Demo()
        .padding(80)
        .background(MuiTokens.bg)
}

File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiCommandMenu.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Input;
using Microsoft.UI.Xaml.Media;
using Maple.UI.Atoms;
using Windows.System;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Command Menu molecule (unified-component-catalog.md §2.4,
    /// "Command Menu" row: "Searchable command palette", built from
    /// Popover, Input, Icon, Text) — an <see cref="MuiInput"/> filter on top
    /// of an anchored <see cref="MuiPopover"/> result list, filtering by
    /// substring match against each command's label via
    /// <see cref="MuiCommandMenuMath"/>.
    ///
    /// Ports `mui-command-menu.component.ts`: every (re)open starts from a
    /// clean search — the previous query never survives a close.
    /// </summary>
    public sealed class MuiCommandMenu : ContentControl
    {
        public static readonly DependencyProperty IsOpenProperty =
            DependencyProperty.Register(nameof(IsOpen), typeof(bool), typeof(MuiCommandMenu),
                new PropertyMetadata(false, (d, _) => ((MuiCommandMenu)d).OnIsOpenChanged()));

        public static readonly DependencyProperty PlacementProperty =
            DependencyProperty.Register(nameof(Placement), typeof(MuiPopoverPlacement), typeof(MuiCommandMenu),
                new PropertyMetadata(MuiPopoverPlacement.Bottom, (d, e) => ((MuiCommandMenu)d)._popover.Placement = (MuiPopoverPlacement)e.NewValue));

        public static readonly DependencyProperty CommandsProperty =
            DependencyProperty.Register(nameof(Commands), typeof(IReadOnlyList<MuiCommandItem>), typeof(MuiCommandMenu),
                new PropertyMetadata(null, (d, _) => ((MuiCommandMenu)d).RebuildList()));

        public static readonly DependencyProperty PlaceholderProperty =
            DependencyProperty.Register(nameof(Placeholder), typeof(string), typeof(MuiCommandMenu),
                new PropertyMetadata("Type a command", (d, e) => ((MuiCommandMenu)d)._search.Placeholder = (string)e.NewValue));

        public bool IsOpen
        {
            get => (bool)GetValue(IsOpenProperty);
            set => SetValue(IsOpenProperty, value);
        }

        public MuiPopoverPlacement Placement
        {
            get => (MuiPopoverPlacement)GetValue(PlacementProperty);
            set => SetValue(PlacementProperty, value);
        }

        public IReadOnlyList<MuiCommandItem>? Commands
        {
            get => (IReadOnlyList<MuiCommandItem>?)GetValue(CommandsProperty);
            set => SetValue(CommandsProperty, value);
        }

        public string Placeholder
        {
            get => (string)GetValue(PlaceholderProperty);
            set => SetValue(PlaceholderProperty, value);
        }

        public event EventHandler<string>? ItemSelected;
        public event EventHandler? CloseRequested;

        private readonly MuiPopover _popover = new();
        private readonly StackPanel _panel = new() { Orientation = Orientation.Vertical, Spacing = 6, MinWidth = 260 };
        private readonly MuiInput _search = new() { Variant = MuiInputVariant.Search, Placeholder = "Type a command" };
        private readonly StackPanel _list = new() { Orientation = Orientation.Vertical, Spacing = 2 };
        private string _query = string.Empty;
        private int _activeIndex;

        public MuiCommandMenu()
        {
            Content = _popover;
            IsTabStop = false;
            IsHitTestVisible = false;

            _panel.Children.Add(_search);
            _panel.Children.Add(_list);
            _popover.PanelContent = _panel;
            _popover.CloseRequested += (_, _) => CloseRequested?.Invoke(this, EventArgs.Empty);

            _search.TextChanged += (_, text) =>
            {
                _query = text;
                _activeIndex = 0;
                RebuildList();
            };
            _panel.KeyDown += OnKeyDown;

            RebuildList();
        }

        private void OnIsOpenChanged()
        {
            if (IsOpen)
            {
                _query = string.Empty;
                _activeIndex = 0;
                _search.Text = string.Empty;
            }
            _popover.IsOpen = IsOpen;
            RebuildList();
        }

        private IReadOnlyList<MuiCommandItem> Filtered() =>
            MuiCommandMenuMath.Filter(Commands ?? Array.Empty<MuiCommandItem>(), _query);

        private void OnKeyDown(object sender, KeyRoutedEventArgs e)
        {
            var filtered = Filtered();
            var clamped = MuiCommandMenuMath.ClampActiveIndex(_activeIndex, filtered.Count);
            if (e.Key == VirtualKey.Down)
            {
                e.Handled = true;
                if (filtered.Count > 0) SetActive(MuiMenuNavMath.WrapIndex(clamped, 1, filtered.Count));
            }
            else if (e.Key == VirtualKey.Up)
            {
                e.Handled = true;
                if (filtered.Count > 0) SetActive(MuiMenuNavMath.WrapIndex(clamped, -1, filtered.Count));
            }
            else if (e.Key == VirtualKey.Enter)
            {
                e.Handled = true;
                if (clamped >= 0 && clamped < filtered.Count) SelectItem(filtered[clamped]);
            }
        }

        private void SetActive(int index)
        {
            _activeIndex = index;
            RebuildHighlights();
        }

        private void SelectItem(MuiCommandItem item) => ItemSelected?.Invoke(this, item.Id);

        private static Brush R(string key) => (Brush)Application.Current.Resources[key];

        private void RebuildList()
        {
            _list.Children.Clear();
            foreach (var command in Filtered())
                _list.Children.Add(BuildRow(command));
            RebuildHighlights();
        }

        private Border BuildRow(MuiCommandItem command)
        {
            var content = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8 };
            if (!string.IsNullOrEmpty(command.IconName))
                content.Children.Add(new MuiIcon { IconName = command.IconName, Size = MuiIconSize.Sm16 });
            content.Children.Add(new MuiText { Text = command.Label, Variant = MuiTextVariant.RowLabel });
            if (!string.IsNullOrEmpty(command.Shortcut))
                content.Children.Add(new MuiText { Text = command.Shortcut, Variant = MuiTextVariant.ValueChip });

            var row = new Border
            {
                CornerRadius = new CornerRadius(6),
                Padding = new Thickness(10, 7, 10, 7),
                Child = content,
            };
            row.PointerPressed += (_, _) => SelectItem(command);
            return row;
        }

        private void RebuildHighlights()
        {
            var clamped = MuiCommandMenuMath.ClampActiveIndex(_activeIndex, _list.Children.Count);
            for (var i = 0; i < _list.Children.Count; i++)
                if (_list.Children[i] is Border row)
                    row.Background = i == clamped ? R("MapleSurfaceHover") : new SolidColorBrush(Microsoft.UI.Colors.Transparent);
        }
    }
}
